File: src/python/quicksort.py
from __future__ import annotations

import sys


def partition(values: list[int], low: int, high: int) -> int:
    pivot = values[high]
    i = low - 1

    for j in range(low, high):
        if values[j] <= pivot:
            i += 1
            values[i], values[j] = values[j], values[i]

    values[i + 1], values[high] = values[high], values[i + 1]
    return i + 1


def quicksort(values: list[int], low: int = 0, high: int | None = None) -> None:
    if high is None:
        high = len(values) - 1

    if low < high:
        pivot_index = partition(values, low, high)
        quicksort(values, low, pivot_index - 1)
        quicksort(values, pivot_index + 1, high)


def parse_input(text: str) -> list[int]:
    tokens = text.split()
    if not tokens:
        raise ValueError("Informe o tamanho da lista.")

    size = int(tokens[0])
    values = [int(token) for token in tokens[1:]]

    if size < 0:
        raise ValueError("O tamanho da lista nao pode ser negativo.")

    if len(values) != size:
        raise ValueError("A quantidade de elementos nao corresponde ao tamanho informado.")

    return values


def main() -> int:
    try:
        values = parse_input(sys.stdin.read())
    except ValueError as error:
        print(f"Entrada invalida: {error}", file=sys.stderr)
        return 1

    quicksort(values)
    print(" ".join(str(value) for value in values))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/c/quicksort.c
#include <stdio.h>
#include <stdlib.h>

static void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

static int partition(int values[], int low, int high) {
    int pivot = values[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (values[j] <= pivot) {
            i++;
            swap(&values[i], &values[j]);
        }
    }

    swap(&values[i + 1], &values[high]);
    return i + 1;
}

static void quicksort(int values[], int low, int high) {
    if (low < high) {
        int pivot_index = partition(values, low, high);

        quicksort(values, low, pivot_index - 1);
        quicksort(values, pivot_index + 1, high);
    }
}

static void print_values(const int values[], int size) {
    for (int i = 0; i < size; i++) {
        if (i > 0) {
            printf(" ");
        }
        printf("%d", values[i]);
    }
    printf("\n");
}

int main(void) {
    int size;

    if (scanf("%d", &size) != 1 || size < 0) {
        fprintf(stderr, "Entrada invalida. Informe o tamanho da lista.\n");
        return 1;
    }

    int *values = malloc((size_t)size * sizeof(int));
    if (size > 0 && values == NULL) {
        fprintf(stderr, "Erro ao alocar memoria.\n");
        return 1;
    }

    for (int i = 0; i < size; i++) {
        if (scanf("%d", &values[i]) != 1) {
            fprintf(stderr, "Entrada invalida. Informe todos os elementos.\n");
            free(values);
            return 1;
        }
    }

    quicksort(values, 0, size - 1);
    print_values(values, size);

    free(values);
    return 0;
}
